In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline
plt.style.use('ggplot')

In [2]:
path = 'C:/Users/ameydalv/data/Data-Collisions.csv'
data = pd.read_csv(path)
data.head()

C:\Users\ameydalv\AppData\Local\Continuum\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3058: DtypeWarning: Columns (33) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


SEVERITYCODE           X          Y  OBJECTID  INCKEY  COLDETKEY REPORTNO  \
0             2 -122.323148  47.703140         1    1307       1307  3502005   
1             1 -122.347294  47.647172         2   52200      52200  2607959   
2             1 -122.334540  47.607871         3   26700      26700  1482393   
3             1 -122.334803  47.604803         4    1144       1144  3503937   
4             2 -122.306426  47.545739         5   17700      17700  1807429   

    STATUS      ADDRTYPE   INTKEY  ... ROADCOND                LIGHTCOND  \
0  Matched  Intersection  37475.0  ...      Wet                 Daylight   
1  Matched         Block      NaN  ...      Wet  Dark - Street Lights On   
2  Matched         Block      NaN  ...      Dry                 Daylight   
3  Matched         Block      NaN  ...      Dry                 Daylight   
4  Matched  Intersection  34387.0  ...      Wet                 Daylight   

  PEDROWNOTGRNT  SDOTCOLNUM SPEEDING ST_COLCODE  \
0           NaN         NaN      NaN         10   
1           NaN   6354039.0      NaN         11   
2           NaN   4323031.0      NaN         32   
3           NaN         NaN      NaN         23   
4           NaN   4028032.0      NaN         10   

                                          ST_COLDESC  SEGLANEKEY  \
0                                  Entering at angle           0   
1  From same direction - both going straight - bo...           0   
2                             One parked--one moving           0   
3                   From same direction - all others           0   
4                                  Entering at angle           0   

   CROSSWALKKEY  HITPARKEDCAR  
0             0             N  
1             0             N  
2             0             N  
3             0             N  
4             0             N  

[5 rows x 38 columns]

In [3]:
data.describe()

SEVERITYCODE              X              Y       OBJECTID  \
count  194673.000000  189339.000000  189339.000000  194673.000000   
mean        1.298901    -122.330518      47.619543  108479.364930   
std         0.457778       0.029976       0.056157   62649.722558   
min         1.000000    -122.419091      47.495573       1.000000   
25%         1.000000    -122.348673      47.575956   54267.000000   
50%         1.000000    -122.330224      47.615369  106912.000000   
75%         2.000000    -122.311937      47.663664  162272.000000   
max         2.000000    -122.238949      47.734142  219547.000000   

              INCKEY      COLDETKEY         INTKEY  SEVERITYCODE.1  \
count  194673.000000  194673.000000   65070.000000   194673.000000   
mean   141091.456350  141298.811381   37558.450576        1.298901   
std     86634.402737   86986.542110   51745.990273        0.457778   
min      1001.000000    1001.000000   23807.000000        1.000000   
25%     70383.000000   70383.000000   28667.000000        1.000000   
50%    123363.000000  123363.000000   29973.000000        1.000000   
75%    203319.000000  203459.000000   33973.000000        2.000000   
max    331454.000000  332954.000000  757580.000000        2.000000   

         PERSONCOUNT       PEDCOUNT    PEDCYLCOUNT       VEHCOUNT  \
count  194673.000000  194673.000000  194673.000000  194673.000000   
mean        2.444427       0.037139       0.028391       1.920780   
std         1.345929       0.198150       0.167413       0.631047   
min         0.000000       0.000000       0.000000       0.000000   
25%         2.000000       0.000000       0.000000       2.000000   
50%         2.000000       0.000000       0.000000       2.000000   
75%         3.000000       0.000000       0.000000       2.000000   
max        81.000000       6.000000       2.000000      12.000000   

        SDOT_COLCODE    SDOTCOLNUM     SEGLANEKEY  CROSSWALKKEY  
count  194673.000000  1.149360e+05  194673.000000  1.946730e+05  
mean       13.867768  7.972521e+06     269.401114  9.782452e+03  
std         6.868755  2.553533e+06    3315.776055  7.226926e+04  
min         0.000000  1.007024e+06       0.000000  0.000000e+00  
25%        11.000000  6.040015e+06       0.000000  0.000000e+00  
50%        13.000000  8.023022e+06       0.000000  0.000000e+00  
75%        14.000000  1.015501e+07       0.000000  0.000000e+00  
max        69.000000  1.307202e+07  525241.000000  5.239700e+06

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 194673 entries, 0 to 194672
Data columns (total 38 columns):
SEVERITYCODE      194673 non-null int64
X                 189339 non-null float64
Y                 189339 non-null float64
OBJECTID          194673 non-null int64
INCKEY            194673 non-null int64
COLDETKEY         194673 non-null int64
REPORTNO          194673 non-null object
STATUS            194673 non-null object
ADDRTYPE          192747 non-null object
INTKEY            65070 non-null float64
LOCATION          191996 non-null object
EXCEPTRSNCODE     84811 non-null object
EXCEPTRSNDESC     5638 non-null object
SEVERITYCODE.1    194673 non-null int64
SEVERITYDESC      194673 non-null object
COLLISIONTYPE     189769 non-null object
PERSONCOUNT       194673 non-null int64
PEDCOUNT          194673 non-null int64
PEDCYLCOUNT       194673 non-null int64
VEHCOUNT          194673 non-null int64
INCDATE           194673 non-null object
INCDTTM           194673 non-null obje

##### Selecting the data as per our requirement
The reason for choosing the fields mentioned below can be found in the notebook - https://github.com/ameyadalvi84/Coursera_Capstone/blob/master/Capstone%20Project.ipynb

In [5]:
df_intermediate = data[['ADDRTYPE','COLLISIONTYPE','PERSONCOUNT','VEHCOUNT','WEATHER','ROADCOND','LIGHTCOND','SEVERITYCODE']]
df_intermediate.head(10)

ADDRTYPE COLLISIONTYPE  PERSONCOUNT  VEHCOUNT   WEATHER ROADCOND  \
0  Intersection        Angles            2         2  Overcast      Wet   
1         Block     Sideswipe            2         2   Raining      Wet   
2         Block    Parked Car            4         3  Overcast      Dry   
3         Block         Other            3         3     Clear      Dry   
4  Intersection        Angles            2         2   Raining      Wet   
5  Intersection        Angles            2         2     Clear      Dry   
6  Intersection        Angles            2         2   Raining      Wet   
7  Intersection        Cycles            3         1     Clear      Dry   
8         Block    Parked Car            2         2     Clear      Dry   
9  Intersection        Angles            2         2     Clear      Dry   

                 LIGHTCOND  SEVERITYCODE  
0                 Daylight             2  
1  Dark - Street Lights On             1  
2                 Daylight             1  
3                 Daylight             1  
4                 Daylight             2  
5                 Daylight             1  
6                 Daylight             1  
7                 Daylight             2  
8                 Daylight             1  
9                 Daylight             2

In [50]:
one_hot_columns = ['ADDRTYPE','COLLISIONTYPE','WEATHER','ROADCOND','LIGHTCOND']
#df_intermediate.isnull().sum()

We are going to drop rows which have Null records.

In [49]:
df_intermediate.dropna(inplace=True)
df_intermediate.isnull().sum()

C:\Users\ameydalv\AppData\Local\Continuum\anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


ADDRTYPE         0
COLLISIONTYPE    0
PERSONCOUNT      0
VEHCOUNT         0
WEATHER          0
ROADCOND         0
LIGHTCOND        0
SEVERITYCODE     0
dtype: int64

Creating Dummy colums for categorical variables

In [8]:
df_final=pd.get_dummies(df_intermediate, columns=one_hot_columns, drop_first=True)
df_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 187504 entries, 0 to 194672
Data columns (total 40 columns):
PERSONCOUNT                           187504 non-null int64
VEHCOUNT                              187504 non-null int64
SEVERITYCODE                          187504 non-null int64
ADDRTYPE_Block                        187504 non-null uint8
ADDRTYPE_Intersection                 187504 non-null uint8
COLLISIONTYPE_Cycles                  187504 non-null uint8
COLLISIONTYPE_Head On                 187504 non-null uint8
COLLISIONTYPE_Left Turn               187504 non-null uint8
COLLISIONTYPE_Other                   187504 non-null uint8
COLLISIONTYPE_Parked Car              187504 non-null uint8
COLLISIONTYPE_Pedestrian              187504 non-null uint8
COLLISIONTYPE_Rear Ended              187504 non-null uint8
COLLISIONTYPE_Right Turn              187504 non-null uint8
COLLISIONTYPE_Sideswipe               187504 non-null uint8
WEATHER_Clear                         187504 non-

Checking to see if there is any correlation between various parameters

In [9]:
df_final.corr()

PERSONCOUNT  VEHCOUNT  SEVERITYCODE  \
PERSONCOUNT                            1.000000  0.400479      0.128368   
VEHCOUNT                               0.400479  1.000000     -0.081014   
SEVERITYCODE                           0.128368 -0.081014      1.000000   
ADDRTYPE_Block                        -0.060863  0.094486     -0.194986   
ADDRTYPE_Intersection                  0.064465 -0.090458      0.199092   
COLLISIONTYPE_Cycles                  -0.042292 -0.305339      0.214620   
COLLISIONTYPE_Head On                  0.020705  0.029271      0.029179   
COLLISIONTYPE_Left Turn                0.049671  0.040773      0.056126   
COLLISIONTYPE_Other                   -0.180873 -0.353875     -0.036636   
COLLISIONTYPE_Parked Car              -0.094683  0.184087     -0.308820   
COLLISIONTYPE_Pedestrian              -0.040252 -0.326174      0.247025   
COLLISIONTYPE_Rear Ended               0.134991  0.211535      0.130250   
COLLISIONTYPE_Right Turn               0.023756  0.012249     -0.026679   
COLLISIONTYPE_Sideswipe                0.039226  0.058439     -0.121182   
WEATHER_Clear                          0.033269  0.014220      0.052011   
WEATHER_Fog/Smog/Smoke                -0.006663 -0.007216      0.003232   
WEATHER_Other                         -0.018536 -0.001872     -0.022498   
WEATHER_Overcast                       0.005787 -0.002999      0.011687   
WEATHER_Partly Cloudy                  0.000518  0.000272      0.003333   
WEATHER_Raining                        0.017402 -0.016192      0.034576   
WEATHER_Severe Crosswind              -0.002237 -0.005172     -0.000585   
WEATHER_Sleet/Hail/Freezing Rain      -0.000949 -0.003395     -0.003310   
WEATHER_Snowing                       -0.005192  0.000470     -0.017624   
WEATHER_Unknown                       -0.087479  0.003157     -0.153030   
ROADCOND_Ice                          -0.014580 -0.015939     -0.013508   
ROADCOND_Oil                          -0.005024 -0.008317      0.002882   
ROADCOND_Other                        -0.005717 -0.009400      0.001982   
ROADCOND_Sand/Mud/Dirt                -0.006925 -0.012009     -0.000083   
ROADCOND_Snow/Slush                   -0.003071 -0.003124     -0.021730   
ROADCOND_Standing Water               -0.008617 -0.014396     -0.002226   
ROADCOND_Unknown                      -0.087567  0.002925     -0.155287   
ROADCOND_Wet                           0.016190 -0.017172      0.036681   
LIGHTCOND_Dark - Street Lights Off    -0.006475 -0.006753     -0.006460   
LIGHTCOND_Dark - Street Lights On      0.010185 -0.034848     -0.005629   
LIGHTCOND_Dark - Unknown Lighting     -0.003634 -0.007065      0.001005   
LIGHTCOND_Dawn                        -0.021519 -0.016210      0.006870   
LIGHTCOND_Daylight                     0.040155  0.036493      0.081099   
LIGHTCOND_Dusk                         0.009189 -0.000980      0.010959   
LIGHTCOND_Other                       -0.009159 -0.005294     -0.005622   
LIGHTCOND_Unknown                     -0.083051  0.005491     -0.149772   

                                    ADDRTYPE_Block  ADDRTYPE_Intersection  \
PERSONCOUNT                              -0.060863               0.064465   
VEHCOUNT                                  0.094486              -0.090458   
SEVERITYCODE                             -0.194986               0.199092   
ADDRTYPE_Block                            1.000000              -0.991226   
ADDRTYPE_Intersection                    -0.991226               1.000000   
COLLISIONTYPE_Cycles                     -0.084201               0.085339   
COLLISIONTYPE_Head On                     0.026672              -0.026315   
COLLISIONTYPE_Left Turn                  -0.297084               0.300265   
COLLISIONTYPE_Other                       0.135950              -0.142851   
COLLISIONTYPE_Parked Car                  0.373216              -0.377924   
COLLISIONTYPE_Pedestrian                 -0.151222               0.150979   
COLLISIONTYPE_Rear Ended            

Splitting the data into train and test data

In [10]:
from sklearn.model_selection import train_test_split
from sklearn.utils import resample

In [11]:
train,test = train_test_split(df_final,test_size = 0.15,random_state = 140)

In [12]:
print(train.shape)
print(test.shape)

(159378, 40)
(28126, 40)


Need to balance the skewed data, by resampling the Severity =1 dataset.

In [13]:
bal = train['SEVERITYCODE'].value_counts()
bal[2]

48349

In [14]:
train_bal_1 = resample(train[train['SEVERITYCODE']==1],
                  replace=False,
                      n_samples = bal[2],
                      random_state=140)
train_bal_1.head()

PERSONCOUNT  VEHCOUNT  SEVERITYCODE  ADDRTYPE_Block  \
92259             3         2             1               1   
137786            2         1             1               1   
140581            2         2             1               0   
104384            3         2             1               1   
87800             2         2             1               1   

        ADDRTYPE_Intersection  COLLISIONTYPE_Cycles  COLLISIONTYPE_Head On  \
92259                       0                     0                      0   
137786                      0                     0                      0   
140581                      1                     0                      0   
104384                      0                     0                      0   
87800                       0                     0                      0   

        COLLISIONTYPE_Left Turn  COLLISIONTYPE_Other  \
92259                         0                    0   
137786                        0                    1   
140581                        1                    0   
104384                        0                    0   
87800                         0                    0   

        COLLISIONTYPE_Parked Car  ...  ROADCOND_Unknown  ROADCOND_Wet  \
92259                          0  ...                 0             0   
137786                         0  ...                 0             0   
140581                         0  ...                 0             0   
104384                         0  ...                 0             0   
87800                          0  ...                 0             0   

        LIGHTCOND_Dark - Street Lights Off  LIGHTCOND_Dark - Street Lights On  \
92259                                    0                                  1   
137786                                   0                                  0   
140581                                   0                                  0   
104384                                   0                                  0   
87800                                    0                                  0   

        LIGHTCOND_Dark - Unknown Lighting  LIGHTCOND_Dawn  LIGHTCOND_Daylight  \
92259                                   0               0                   0   
137786                                  0               0                   1   
140581                                  0               0                   1   
104384                                  0               0                   1   
87800                                   0               0                   1   

        LIGHTCOND_Dusk  LIGHTCOND_Other  LIGHTCOND_Unknown  
92259                0                0                  0  
137786               0                0                  0  
140581               0                0                  0  
104384               0                0                  0  
87800                0                0                  0  

[5 rows x 40 columns]

In [15]:
train_bal_2 = pd.DataFrame(train[train['SEVERITYCODE']==2])
train_bal_2.shape

(48349, 40)

Creating the dataset with equal count of data for Severity 1 and Severity 2.

In [16]:
downsample_train = pd.concat([train_bal_1,train_bal_2])
downsample_train.shape

(96698, 40)

Normalizing the data for Person Count and Vehical Count

In [17]:
from sklearn.preprocessing import StandardScaler

scale1 = StandardScaler(copy=False)
scale2 = StandardScaler(copy=False)
scale1.fit(downsample_train[['PERSONCOUNT']])
downsample_train['PerCount'] = scale1.transform(downsample_train[['PERSONCOUNT']])
scale2.fit(downsample_train[['VEHCOUNT']])
downsample_train['VehCount'] = scale2.transform(downsample_train[['VEHCOUNT']])
downsample_train.head()

PERSONCOUNT  VEHCOUNT  SEVERITYCODE  ADDRTYPE_Block  \
92259             3         2             1               1   
137786            2         1             1               1   
140581            2         2             1               0   
104384            3         2             1               1   
87800             2         2             1               1   

        ADDRTYPE_Intersection  COLLISIONTYPE_Cycles  COLLISIONTYPE_Head On  \
92259                       0                     0                      0   
137786                      0                     0                      0   
140581                      1                     0                      0   
104384                      0                     0                      0   
87800                       0                     0                      0   

        COLLISIONTYPE_Left Turn  COLLISIONTYPE_Other  \
92259                         0                    0   
137786                        0                    1   
140581                        1                    0   
104384                        0                    0   
87800                         0                    0   

        COLLISIONTYPE_Parked Car  ...  LIGHTCOND_Dark - Street Lights Off  \
92259                          0  ...                                   0   
137786                         0  ...                                   0   
140581                         0  ...                                   0   
104384                         0  ...                                   0   
87800                          0  ...                                   0   

        LIGHTCOND_Dark - Street Lights On  LIGHTCOND_Dark - Unknown Lighting  \
92259                                   1                                  0   
137786                                  0                                  0   
140581                                  0                                  0   
104384                                  0                                  0   
87800                                   0                                  0   

        LIGHTCOND_Dawn  LIGHTCOND_Daylight  LIGHTCOND_Dusk  LIGHTCOND_Other  \
92259                0                   0               0                0   
137786               0                   1               0                0   
140581               0                   1               0                0   
104384               0                   1               0                0   
87800                0                   1               0                0   

        LIGHTCOND_Unknown  PerCount  VehCount  
92259                   0  0.324621  0.080431  
137786                  0 -0.371995 -1.561434  
140581                  0 -0.371995  0.080431  
104384                  0  0.324621  0.080431  
87800                   0 -0.371995  0.080431  

[5 rows x 42 columns]

In [18]:
X_train = downsample_train.drop(['SEVERITYCODE','PERSONCOUNT','VEHCOUNT'],axis=1)
y_train = downsample_train['SEVERITYCODE']

In [19]:
test['PerCount'] = scale1.transform(test[['PERSONCOUNT']])
test['VehCount'] = scale2.transform(test[['VEHCOUNT']])
#test.head()
X_test = test.drop(['SEVERITYCODE','PERSONCOUNT','VEHCOUNT'],axis=1)
y_test = test['SEVERITYCODE']

C:\Users\ameydalv\AppData\Local\Continuum\anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
C:\Users\ameydalv\AppData\Local\Continuum\anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [20]:
assert X_train.shape[0] == y_train.shape[0]
assert X_test.shape[0] == y_test.shape[0]

### Model1 - Logistic Regression

In [21]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import GridSearchCV
clf = LogisticRegression()

In [25]:
clf.fit(X_train,y_train)

C:\Users\ameydalv\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='warn', n_jobs=None, penalty='l2',
                   random_state=None, solver='warn', tol=0.0001, verbose=0,
                   warm_start=False)

In [26]:
y_pred = clf.predict(X_test)

In [22]:
from sklearn.metrics import confusion_matrix, classification_report

Checking the f1 score for the model

In [28]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           1       0.87      0.60      0.71     19605
           2       0.47      0.80      0.59      8521

    accuracy                           0.66     28126
   macro avg       0.67      0.70      0.65     28126
weighted avg       0.75      0.66      0.67     28126



Using GridSearch to search best parameters

In [40]:
parameters = {'C':(1,1e5,1e2),
    'solver':('newton-cg', 'lbfgs', 'sag')}
grid_clf = GridSearchCV(LogisticRegression(penalty = 'l2',n_jobs=-1),param_grid=parameters)

In [41]:
grid_clf.fit(X_train,y_train)

C:\Users\ameydalv\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\model_selection\_split.py:1978: FutureWarning: The default value of cv will change from 3 to 5 in version 0.22. Specify it explicitly to silence this warning.
  warnings.warn(CV_WARNING, FutureWarning)
C:\Users\ameydalv\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\linear_model\sag.py:337: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)


GridSearchCV(cv='warn', error_score='raise-deprecating',
             estimator=LogisticRegression(C=1.0, class_weight=None, dual=False,
                                          fit_intercept=True,
                                          intercept_scaling=1, l1_ratio=None,
                                          max_iter=100, multi_class='warn',
                                          n_jobs=-1, penalty='l2',
                                          random_state=None, solver='warn',
                                          tol=0.0001, verbose=0,
                                          warm_start=False),
             iid='warn', n_jobs=None,
             param_grid={'C': (1, 100000.0, 100.0),
                         'solver': ('newton-cg', 'lbfgs', 'sag')},
             pre_dispatch='2*n_jobs', refit=True, return_train_score=False,
             scoring=None, verbose=0)

In [42]:
print(grid_clf.best_params_)
print(grid_clf.best_score_)

{'C': 100.0, 'solver': 'sag'}
0.702113797596641


In [36]:
parameters1 = {'C':(1,1e5,1e2),
              'solver': ('liblinear','saga')}

grid_clf1 = GridSearchCV(LogisticRegression(penalty='l1',random_state = 140,n_jobs=-1),param_grid=parameters1)

In [38]:
#grid_clf1.fit(X_train,y_train)
print(grid_clf1.best_params_)
print(grid_clf1.best_score_)

{'C': 1, 'solver': 'saga'}
0.7020620902190324


In [45]:
y_pred = grid_clf1.predict(X_test)
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           1       0.87      0.60      0.71     19605
           2       0.47      0.80      0.59      8521

    accuracy                           0.66     28126
   macro avg       0.67      0.70      0.65     28126
weighted avg       0.75      0.66      0.67     28126



In [46]:
y_pred = grid_clf.predict(X_test)
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           1       0.87      0.60      0.71     19605
           2       0.47      0.80      0.59      8521

    accuracy                           0.66     28126
   macro avg       0.67      0.70      0.65     28126
weighted avg       0.75      0.66      0.67     28126



### Model2 - SVM

SVM is not working and not giving any output.

In [26]:
from sklearn.svm import SVC
import time

In [25]:
svc_clf = SVC(gamma='auto',C=1e2,random_state=140)

In [ ]:
start_time = time.time()
svc_clf.fit(X_train,y_train)
print(time.time()-start_time)

In [ ]:
y_pred = svc_clf.predict(X_test)
print(classification_report(y_test,y_pred))

In [ ]:
parameters = {'kernel':('rfb','sigmoid','linear'),
             'gamma':('auto','scale')}
grid_clf=GridSearchCV(svc_clf,param=parameters)

### Model3 - Random Forest

In [39]:
from sklearn.ensemble import RandomForestClassifier
import time

In [40]:
parameters = {'n_estimators':[499,500,501]}
             #'criterion':('gini'),
             #'max_depth':[8,10]}
            # 'max_features':('auto','sqrt')}

In [41]:
rf_clf = RandomForestClassifier(criterion='gini',n_jobs=-1,random_state=140,max_features='auto',max_depth=10)

In [42]:
#start_time = time.time()
rf_clf.fit(X_train,y_train)
#print(time.time()-start_time)

C:\Users\ameydalv\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\ensemble\forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


0.5524294376373291


Using GridSearch to get the best parameters

In [43]:
grid_clf=GridSearchCV(rf_clf,param_grid=parameters,verbose=10,n_jobs=-1)

In [44]:
start_time = time.time()
grid_clf.fit(X_train,y_train)
print(time.time()-start_time)

C:\Users\ameydalv\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\model_selection\_split.py:1978: FutureWarning: The default value of cv will change from 3 to 5 in version 0.22. Specify it explicitly to silence this warning.
  warnings.warn(CV_WARNING, FutureWarning)
[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.


Fitting 3 folds for each of 3 candidates, totalling 9 fits


[Parallel(n_jobs=-1)]: Done   2 out of   9 | elapsed:  1.5min remaining:  5.4min
[Parallel(n_jobs=-1)]: Done   3 out of   9 | elapsed:  1.5min remaining:  3.1min
[Parallel(n_jobs=-1)]: Done   4 out of   9 | elapsed:  1.5min remaining:  1.9min
[Parallel(n_jobs=-1)]: Done   5 out of   9 | elapsed:  1.5min remaining:  1.2min
[Parallel(n_jobs=-1)]: Done   6 out of   9 | elapsed:  1.6min remaining:   46.9s
[Parallel(n_jobs=-1)]: Done   7 out of   9 | elapsed:  1.6min remaining:   26.8s
[Parallel(n_jobs=-1)]: Done   9 out of   9 | elapsed:  1.7min remaining:    0.0s
[Parallel(n_jobs=-1)]: Done   9 out of   9 | elapsed:  1.7min finished


120.19728660583496


In [45]:
print(grid_clf.best_params_)
print(grid_clf.best_score_)

{'n_estimators': 500}
0.7103973194895448


Saving the model

In [47]:
import pickle
pickle.dump(grid_clf,open('rfm_model_save.sav','wb'))

Checking the f1 score

In [48]:
y_pred = grid_clf.predict(X_test)
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           1       0.88      0.60      0.72     19605
           2       0.47      0.81      0.60      8521

    accuracy                           0.67     28126
   macro avg       0.67      0.71      0.66     28126
weighted avg       0.76      0.67      0.68     28126



We got accuracy of **67%** and score of **0.71** for following paramaters
<br>criterion='gini'</br>
<br>max_features='auto'</br
<br>max_depth=10</br>
<br>n_estimators = 500</br>

### Model 4 - Xgboost Classifer

In [78]:
!pip install xgboost

In [23]:
from xgboost import XGBClassifier

In [36]:
xgb_clf = XGBClassifier(n_estimators = 500,objective="binary:logistic", random_state=140,max_depth = 10,
                       n_jobs = -1,
                       booster = 'gbtree' 
                       )

In [37]:
start_time = time.time()
xgb_clf.fit(X_train,y_train)
print(time.time()-start_time)

56.71140217781067


In [38]:
y_pred = xgb_clf.predict(X_test)
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           1       0.87      0.62      0.72     19605
           2       0.47      0.78      0.59      8521

    accuracy                           0.67     28126
   macro avg       0.67      0.70      0.66     28126
weighted avg       0.75      0.67      0.68     28126



In [ ]:
parameter = {'n_estimators':[10,100,1000],
             'max_depth':[1,10,100],
             'learning_rate':[0.001,0.01,0.1],
             'booster':('gbtree','dart')       
}

In [30]:
from sklearn.model_selection import RandomizedSearchCV

rv_clf = RandomizedSearchCV(xgb_clf,param_distributions=parameter,n_iter=200,verbose=1,n_jobs=2,return_train_score=True)

In [ ]:
rv_clf.fit(X_train,y_train)

C:\Users\ameydalv\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\model_selection\_split.py:1978: FutureWarning: The default value of cv will change from 3 to 5 in version 0.22. Specify it explicitly to silence this warning.
  warnings.warn(CV_WARNING, FutureWarning)
C:\Users\ameydalv\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\model_selection\_search.py:266: UserWarning: The total space of parameters 54 is smaller than n_iter=200. Running 54 iterations. For exhaustive searches, use GridSearchCV.
  % (grid_size, self.n_iter, grid_size), UserWarning)
[Parallel(n_jobs=2)]: Using backend LokyBackend with 2 concurrent workers.


Fitting 3 folds for each of 54 candidates, totalling 162 fits


[Parallel(n_jobs=2)]: Done  46 tasks      | elapsed: 32.8min


### Conclusion

With XGBoost and Random Forest classifier i am getting an model with accuracy of 0.67.
I could not use grid search with SVM or XGBoost as it consumed lot of resources and resulted in system crashing.
Hence for me the best model is Random Forest model with following paramters.

<ol><li>criterion='gini'</li>
<li>max_features='auto'</li>
<li>max_depth=10</li>
<li>n_estimators = 500</li></ol>